streampark-console/streampark-console-service/src/main/scala/org/apache/streampark/console/core/task/FlinkK8sChangeListener.scala

```diff
@@ -0,0 +1,222 @@
+              setByJobStatusCV(app, jobSnap)
+
+              // send alert
+              val state: FlinkAppState = FlinkAppState.of(app.getState)
+              if (
+                FlinkAppState.FAILED == state || FlinkAppState.LOST == state || FlinkAppState.RESTARTING == state || FlinkAppState.FINISHED == state
+              ) {
+                executor.execute(() => alertService.alert(app, state))
+              }
+
+            }
+            .retryN(3)
+            .ignore
+      }
+  }
+
+  def subscribeMetricsChange: UIO[Unit] = {
+    FlinkK8sObserver.clusterMetricsSnaps
+      .flatSubscribe()
+      .foreach {
+        metricsSnap =>
+          ZIO
+            .attempt {
+              val namespaceAndName: (Namespace, Name) = metricsSnap._1
+              val trackKey: ZIO[Any, TrackKeyNotFound, TrackKey] = FlinkK8sObserver.trackedKeys
+                .find(
+                  trackedKey =>
+                    trackedKey.clusterNamespace == namespaceAndName._1 && trackedKey.clusterName == namespaceAndName._2)
+                .someOrFail(TrackKeyNotFound(namespaceAndName._1, namespaceAndName._2))
+
+              val app: Application = applicationService.getById(trackKey.map(_.id))
+
+              // discard session mode change
+              if (app == null || ExecutionMode.isKubernetesSessionMode(app.getExecutionMode))
+                return ZIO.unit
+
+              val clusterMetrics: ClusterMetrics = metricsSnap._2
+              app.setJmMemory(clusterMetrics.totalJmMemory)
+              app.setTmMemory(clusterMetrics.totalTmMemory)
+              app.setTotalTM(clusterMetrics.totalTm)
+              app.setTotalSlot(clusterMetrics.totalSlot)
+              app.setAvailableSlot(clusterMetrics.availableSlot)
+              applicationService.persistMetrics(app)
+            }
+            .retryN(3)
+            .ignore
+      }
+  }
+
+  def subscribeRestSvcEndpointChange: UIO[Unit] = {
+    FlinkK8sObserver.restSvcEndpointSnaps
+      .flatSubscribe()
+      .foreach {
+        restSvcEndpointSnap =>
+          ZIO
+            .attempt {
+
+              val namespaceAndName: (Namespace, Name) = restSvcEndpointSnap._1
+              val trackKey: ZIO[Any, TrackKeyNotFound, TrackKey] = FlinkK8sObserver.trackedKeys
+                .find(
+                  trackedKey =>
+                    trackedKey.clusterNamespace == namespaceAndName._1 && trackedKey.clusterName == namespaceAndName._2)
+                .someOrFail(TrackKeyNotFound(namespaceAndName._1, namespaceAndName._2))
+              val restSvcEndpoint: RestSvcEndpoint = restSvcEndpointSnap._2
+
+              val app: Application = applicationService.getById(trackKey.map(_.id))
+
+              val flinkCluster: FlinkCluster = flinkClusterService.getById(app.getFlinkClusterId)
+
+              if (restSvcEndpoint == null || restSvcEndpoint.chooseRest == null) return ZIO.unit
+              val url = restSvcEndpoint.chooseRest
```

Comment:
Trends to use restSvcEndpoint.ipRest here

Reply:
Ok, thanks for reviewing the code